sw/airborne/modules/px4io_flash/px4io_flash.c

```diff
@@ -0,0 +1,276 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/px4io_flash/px4io_flash.c"
+ * @author Kevin van Hecke
+ * Flashes the px4io f1 through the px4 bootloader.
+ * Assumes the telem2 port on the Pixhawk is connected to a ttyACM device (blackmagic probe)
+ */
+
+#include "modules/px4io_flash/px4io_flash.h"
+//#include "subsystems/datalink/downlink.h"
+#include "modules/px4io_flash/protocol.h"
+#include "mcu_periph/sys_time_arch.h"
+#include "subsystems/intermcu/intermcu_ap.h"
+
+// Serial Port
+#include "mcu_periph/uart.h"
+#include "mcu_periph/usb_serial.h"
+
+#include "led.h"
+
+// define coms link for px4io f1
+#define PX4IO_PORT   (&((PX4IO_UART).device))
+#define TELEM2_PORT   (&((TELEM2_UART).device))
+
+// weird that these below are not in protocol.h, which is from the firmware px4 repo
+// below is copied from qgroundcontrol:
+#define PROTO_INSYNC            0x12   ///< 'in sync' byte sent before status
+#define PROTO_EOC               0x20   ///< end of command
+// Reply bytes
+#define PROTO_OK                0x10   ///< INSYNC/OK      - 'ok' response
+#define PROTO_FAILED            0x11   ///< INSYNC/FAILED  - 'fail' response
+#define PROTO_INVALID           0x13 ///< INSYNC/INVALID - 'invalid' response for bad commands
+// Command bytes
+#define PROTO_GET_SYNC          0x21   ///< NOP for re-establishing sync
+#define PROTO_GET_DEVICE        0x22   ///< get device ID bytes
+#define PROTO_CHIP_ERASE        0x23   ///< erase program area and reset program address
+#define PROTO_LOAD_ADDRESS      0x24 ///< set next programming address
+#define PROTO_PROG_MULTI        0x27   ///< write bytes at program address and increment
+#define PROTO_GET_CRC           0x29 ///< compute & return a CRC
+#define PROTO_BOOT              0x30   ///< boot the application
+
+bool_t setToBootloaderMode;
+
+void px4ioflash_init(void)
+{
+  setToBootloaderMode = FALSE;
+}
+
+void px4ioflash_event(void)
+{
+  // setToBootloaderMode=true;
+  if (PX4IO_PORT->char_available(PX4IO_PORT->periph)) {
+    if (!setToBootloaderMode) {
+      //ignore anything coming from IO if not in bootloader mode (which should be nothing)
+    } else {
+      //relay everything from IO to the laptop
+      unsigned char b = PX4IO_PORT->get_byte(PX4IO_PORT->periph);
+      TELEM2_PORT->put_byte(TELEM2_PORT->periph, b);
+      TELEM2_PORT->send_message(TELEM2_PORT->periph);
```

Comment:
I managed to get FBW flashing over USB working... but currently I need a hack to do so: 4f220522862c0b830b5267d81c20378115afe7a6
In any case I've added automatic sending of the tx fifo if anything is in there from the event function, I also tried to make sure that manually calling the send_message function is not impacted by this change. However, chances are this fails due to the same reason I need the timeout hack. 

For some reason the usbd_ep_stall_get function does not seem to function properly always... Internally it checks the USB_EP_TX_STAT bit, which seems about right. No clue on how to further debug this. Any ideas?